File: implementations/csharp/protocol/src/ShuaiTunnel.Protocol/Codec/PacketCodec.cs
using System.Buffers.Binary;
using System.Collections.Frozen;
using System.Text;
using System.Text.Json.Nodes;
using ShuaiTunnel.Protocol.Packets;

namespace ShuaiTunnel.Protocol.Codec;

/// <summary>
/// Strict control protocol v2 codec. The frame header is
/// magic(4), version(1), serializer(1), command(1), bodyLength(4), all integers big-endian.
/// CompactBinary (wire id 4) is the sole serializer. NAT data is never compressed.
/// </summary>
public static class PacketCodec
{
    public const int MagicNumber = 0x14353565;
    public const byte ProtocolVersion = 2;
    public const int HeaderSize = 11;
    public const int MaxFrameSize = 32 * 1024 * 1024;
    public const int PreAuthMaxFrameSize = 16 * 1024;
    public const int MaxMessageBodySize = 1024 * 1024;
    public const int MaxNatMetadataSize = ushort.MaxValue;

    public const int NatBodyHeaderSize = 16;
    public const byte NatFlagEndStream = 1;

    private static readonly FrozenDictionary<sbyte, Func<byte[], Packet>> CommandToDecoder =
        new Dictionary<sbyte, Func<byte[], Packet>>
        {
            [Command.LoginRequest] = CompactBinarySerializer.Deserialize<LoginRequestPacket>,
            [Command.LoginResponse] = CompactBinarySerializer.Deserialize<LoginResponsePacket>,
            [Command.MessageRequest] = CompactBinarySerializer.Deserialize<MessageRequestPacket>,
            [Command.MessageResponse] = CompactBinarySerializer.Deserialize<MessageResponsePacket>,
            [Command.LogoutRequest] = CompactBinarySerializer.Deserialize<LogoutRequestPacket>,
            [Command.LogoutResponse] = CompactBinarySerializer.Deserialize<LogoutResponsePacket>,
            [Command.HeartbeatRequest] = CompactBinarySerializer.Deserialize<HeartbeatRequestPacket>,
            [Command.HeartbeatResponse] = CompactBinarySerializer.Deserialize<HeartbeatResponsePacket>,
        }.ToFrozenDictionary();

    public static byte[] Encode(Packet packet)
    {
        ArgumentNullException.ThrowIfNull(packet);
        if (packet.Version != ProtocolVersion)
        {
            throw new InvalidDataException($"unsupported protocol version: {packet.Version}");
        }
        if (!IsKnownCommand(packet.Command))
        {
            throw new InvalidDataException($"unknown command byte: {packet.Command}");
        }

        var body = packet is NatMessagePacket nat
            ? EncodeNatBody(nat)
            : CompactBinarySerializer.Serialize(packet);
        ValidateBodyLength(packet.Command, body.Length);

        var result = new byte[HeaderSize + body.Length];
        BinaryPrimitives.WriteInt32BigEndian(result.AsSpan(0, 4), MagicNumber);
        result[4] = ProtocolVersion;
        result[5] = SerializerAlgorithm.CompactBinary;
        result[6] = unchecked((byte)packet.Command);
        BinaryPrimitives.WriteInt32BigEndian(result.AsSpan(7, 4), body.Length);
        body.CopyTo(result.AsSpan(HeaderSize));
        return result;
    }

    /// <summary>
    /// Decodes the first complete frame in <paramref name="input"/>. Additional bytes belong to
    /// the next stream frame and are reported through <paramref name="consumed"/>.
    /// </summary>
    public static bool TryDecode(ReadOnlySpan<byte> input, out Packet? packet, out int consumed)
    {
        packet = null;
        consumed = 0;
        if (input.Length < HeaderSize)
        {
            return false;
        }
        var (command, length) = DecodeHeader(input[..HeaderSize]);
        if (input.Length < HeaderSize + length)
        {
            return false;
        }

        var body = input.Slice(HeaderSize, length).ToArray();
        try
        {
            packet = command == Command.NatMessage
                ? DecodeNatBody(body)
                : CommandToDecoder[command](body);
        }
        catch (InvalidDataException)
        {
            throw;
        }
        catch (Exception exception)
        {
            throw new InvalidDataException($"malformed body for command {command}", exception);
        }
        packet.Version = ProtocolVersion;
        consumed = HeaderSize + length;
        return true;
    }

    /// <summary>Decodes exactly one frame and rejects truncation or trailing bytes.</summary>
    public static Packet DecodeExact(ReadOnlySpan<byte> input)
    {
        if (!TryDecode(input, out var packet, out var consumed))
        {
            throw new InvalidDataException("frame is truncated");
        }
        if (consumed != input.Length)
        {
            throw new InvalidDataException("frame has trailing bytes");
        }
        return packet ?? throw new InvalidDataException("frame did not contain a packet");
    }

    public static (sbyte Command, int BodyLength) DecodeHeader(ReadOnlySpan<byte> header)
    {
        if (header.Length < HeaderSize)
        {
            throw new InvalidDataException("frame header is truncated");
        }
        var magic = BinaryPrimitives.ReadInt32BigEndian(header[..4]);
        if (magic != MagicNumber)
        {
            throw new InvalidDataException($"bad magic: 0x{magic:X8}");
        }
        if (header[4] != ProtocolVersion)
        {
            throw new InvalidDataException($"unsupported protocol version: {header[4]}");
        }
        if (header[5] != SerializerAlgorithm.CompactBinary)
        {
            throw new InvalidDataException($"unsupported serializer: {header[5]}");
        }
        var command = unchecked((sbyte)header[6]);
        if (!IsKnownCommand(command))
        {
            throw new InvalidDataException($"unknown command byte: {command}");
        }
        var length = BinaryPrimitives.ReadInt32BigEndian(header.Slice(7, 4));
        ValidateBodyLength(command, length);
        return (command, length);
    }

    public static void ValidateBodyLength(sbyte command, int length)
    {
        var maximum = command switch
        {
            Command.LoginRequest or Command.LoginResponse => PreAuthMaxFrameSize - HeaderSize,
            Command.MessageRequest or Command.MessageResponse => MaxMessageBodySize,
            _ => MaxFrameSize - HeaderSize,
        };
        if (length < 0 || length > maximum)
        {
            throw new InvalidDataException($"command {command} body exceeds limit: {length}/{maximum}");
        }
    }

    private static byte[] EncodeNatBody(NatMessagePacket nat)
    {
        if (NatMessageTypeExtensions.FromCode(nat.NatMessageType.Code()) is null)
        {
            throw new InvalidDataException($"unknown NAT message type: {(int)nat.NatMessageType}");
        }
        var metadata = SerializeMeta(nat.MetaData ?? new Dictionary<string, object?>());
        if (nat.MetaData is null or { Count: 0 })
        {
            metadata = Array.Empty<byte>();
        }
        if (metadata.Length > MaxNatMetadataSize)
        {
            throw new InvalidDataException("NAT metadata exceeds limit");
        }
        var data = nat.Data ?? Array.Empty<byte>();
        ValidateNatSemantics(nat.NatMessageType, nat.Flags, nat.StreamId, nat.Value, metadata.Length, data.Length);
        if ((long)NatBodyHeaderSize + metadata.Length + data.Length > MaxFrameSize - HeaderSize)
        {
            throw new InvalidDataException("NAT body exceeds frame limit");
        }
        var body = new byte[NatBodyHeaderSize + metadata.Length + data.Length];
        body[0] = checked((byte)nat.NatMessageType.Code());
        body[1] = nat.Flags;
        BinaryPrimitives.WriteUInt16BigEndian(body.AsSpan(2, 2), checked((ushort)metadata.Length));
        BinaryPrimitives.WriteUInt32BigEndian(body.AsSpan(4, 4), nat.StreamId);
        BinaryPrimitives.WriteUInt32BigEndian(body.AsSpan(8, 4), nat.Value);
        BinaryPrimitives.WriteUInt32BigEndian(body.AsSpan(12, 4), checked((uint)data.Length));
        metadata.CopyTo(body.AsSpan(NatBodyHeaderSize));
        data.CopyTo(body.AsSpan(NatBodyHeaderSize + metadata.Length));
        return body;
    }

    private static NatMessagePacket DecodeNatBody(ReadOnlySpan<byte> body)
    {
        if (body.Length < NatBodyHeaderSize)
        {
            throw new InvalidDataException("NAT_MESSAGE body too short");
        }
        var natType = NatMessageTypeExtensions.FromCode(body[0])
            ?? throw new InvalidDataException($"unknown NAT message type: {body[0]}");
        var flags = body[1];
        if ((flags & ~NatFlagEndStream) != 0)
        {
            throw new InvalidDataException($"unknown NAT flags: {flags}");
        }
        var metadataLength = BinaryPrimitives.ReadUInt16BigEndian(body.Slice(2, 2));
        var streamId = BinaryPrimitives.ReadUInt32BigEndian(body.Slice(4, 4));
        var value = BinaryPrimitives.ReadUInt32BigEndian(body.Slice(8, 4));
        var dataLength = BinaryPrimitives.ReadUInt32BigEndian(body.Slice(12, 4));
        var expectedLength = (long)NatBodyHeaderSize + metadataLength + dataLength;
        if (expectedLength != body.Length)
        {
            throw new InvalidDataException("NAT_MESSAGE metadata/data length mismatch");
        }
        var metadata = metadataLength == 0
            ? new Dictionary<string, object?>()
            : DeserializeMeta(body.Slice(NatBodyHeaderSize, metadataLength))
                ?? throw new InvalidDataException("NAT metadata must be a JSON object");
        var data = dataLength == 0
            ? null
            : body.Slice(NatBodyHeaderSize + metadataLength, checked((int)dataLength)).ToArray();
        ValidateNatSemantics(natType, flags, streamId, value, metadataLength, checked((int)dataLength));
        return new NatMessagePacket
        {
            NatMessageType = natType,
            Flags = flags,
            StreamId = streamId,
            Value = value,
            MetaData = metadata,
            Data = data,
        };
    }

    private static void ValidateNatSemantics(
        NatMessageType type,
        byte flags,
        uint streamId,
        uint value,
        int metadataLength,
        int dataLength)
    {
        var streamFrame = type is NatMessageType.Open or NatMessageType.Fin or NatMessageType.Data
            or NatMessageType.Rst or NatMessageType.WindowUpdate;
        if (streamFrame == (streamId == 0))
        {
            throw new InvalidDataException(streamFrame
                ? "stream frame requires a non-zero stream id"
                : "connection frame requires stream id zero");
        }
        if (type != NatMessageType.Data && flags != 0)
        {
            throw new InvalidDataException("flags are only valid on DATA");
        }
        if (type == NatMessageType.Data && (metadataLength != 0 || value != 0))
        {
            throw new InvalidDataException("DATA cannot carry metadata/value");
        }
        if (type == NatMessageType.Fin && (dataLength != 0 || flags != 0))
        {
            throw new InvalidDataException("FIN cannot carry binary data/flags");
        }
        if (type == NatMessageType.WindowUpdate &&
            (metadataLength != 0 || dataLength != 0 || flags != 0))
        {
            throw new InvalidDataException("WINDOW_UPDATE cannot carry payload");
        }
        if (type == NatMessageType.WindowUpdate && value == 0)
        {
            throw new InvalidDataException("WINDOW_UPDATE credit must be positive");
        }
        if (type == NatMessageType.Fin && value != 0)
        {
            throw new InvalidDataException("FIN value must be zero");
        }
        if (type == NatMessageType.Rst && dataLength != 0)
        {
            throw new InvalidDataException("RST cannot carry binary data");
        }
        if (!streamFrame && (value != 0 || flags != 0 || dataLength != 0))
        {
            throw new InvalidDataException("connection control frame cannot carry stream value/data");
        }
    }

    private static bool IsKnownCommand(sbyte command) =>
        command == Command.NatMessage || CommandToDecoder.ContainsKey(command);

    private static byte[] SerializeMeta(Dictionary<string, object?> metaData)
    {
        var node = new JsonObject();
        foreach (var (key, value) in metaData)
        {
            node[key] = ToJsonNode(value);
        }
        return Encoding.UTF8.GetBytes(node.ToJsonString());
    }

    private static JsonNode? ToJsonNode(object? value) => value switch
    {
        null => null,
        string text => JsonValue.Create(text),
        bool boolean => JsonValue.Create(boolean),
        int integer => JsonValue.Create(integer),
        long longValue => JsonValue.Create(longValue),
        double doubleValue => JsonValue.Create(doubleValue),
        float floatValue => JsonValue.Create(floatValue),
        decimal decimalValue => JsonValue.Create(decimalValue),
        JsonNode jsonNode => jsonNode.DeepClone(),
        IEnumerable<string> strings => new JsonArray(strings
            .Select(static item => (JsonNode?)JsonValue.Create(item)).ToArray()),
        IEnumerable<object?> objects => new JsonArray(objects.Select(ToJsonNode).ToArray()),
        IReadOnlyDictionary<string, object?> map => new JsonObject(
            map.Select(static pair => KeyValuePair.Create(pair.Key, ToJsonNode(pair.Value)))),
        _ => JsonValue.Create(value.ToString()),
    };

    private static Dictionary<string, object?>? DeserializeMeta(ReadOnlySpan<byte> json)
    {
        if (json.IsEmpty)
        {
            throw new InvalidDataException("NAT JSON metadata is empty");
        }
        var node = JsonNode.Parse(Encoding.UTF8.GetString(json));
        if (node is not JsonObject obj)
        {
            return null;
        }
        var map = new Dictionary<string, object?>(obj.Count);
        foreach (var (key, value) in obj)
        {
            map[key] = FromJsonNode(value);
        }
        return map;
    }

    private static object? FromJsonNode(JsonNode? node)
    {
        if (node is null)
        {
            return null;
        }
        if (node is JsonValue value)
        {
            if (value.TryGetValue<bool>(out var boolean)) return boolean;
            if (value.TryGetValue<long>(out var integer)) return integer;
            if (value.TryGetValue<double>(out var number)) return number;
            if (value.TryGetValue<string>(out var text)) return text;
        }
        if (node is JsonArray array)
        {
            return array.Select(FromJsonNode).ToList();
        }
        if (node is JsonObject obj)
        {
            return obj.ToDictionary(
                static pair => pair.Key,
                static pair => FromJsonNode(pair.Value),
                StringComparer.Ordinal);
        }
        throw new InvalidDataException("unsupported NAT metadata JSON value");
    }
}

File: implementations/java/common/src/main/java/com/theshuai/common/protocol/PacketCodec.java
package com.theshuai.common.protocol;

import com.fasterxml.jackson.core.type.TypeReference;
import com.theshuai.common.command.Command;
import com.theshuai.common.protocol.request.HeartBeatRequestPacket;
import com.theshuai.common.protocol.request.LoginRequestPacket;
import com.theshuai.common.protocol.request.LogoutRequestPacket;
import com.theshuai.common.protocol.request.MessageRequestPacket;
import com.theshuai.common.protocol.response.HeartBeatResponsePacket;
import com.theshuai.common.protocol.response.LoginResponsePacket;
import com.theshuai.common.protocol.response.LogoutResponsePacket;
import com.theshuai.common.protocol.response.MessageResponsePacket;
import com.theshuai.common.serialize.Serializer;
import com.theshuai.common.serialize.SerializerAlgorithm;
import com.theshuai.common.util.JsonUtil;
import io.netty.buffer.ByteBuf;

import java.util.HashMap;
import java.util.Map;

/**
 * Control protocol v2. The fixed 11-byte frame header remains deliberately simple:
 * magic(4), version(1), serializer(1), command(1), bodyLength(4).
 *
 * <p>Version 2 has one serializer contract: CompactBinary (wire id 4). NAT_MESSAGE keeps a
 * command-specific body, but no longer advertises FastJSON and never wraps specus bytes in a
 * compression envelope.
 */
public final class PacketCodec {
    public static final int MAGIC_NUMBER = 0x14353565;
    public static final byte PROTOCOL_VERSION = 2;
    public static final int HEADER_SIZE = 11;
    public static final int PRE_AUTH_MAX_FRAME_SIZE = 16 * 1024;
    public static final int MAX_NAT_METADATA_BYTES = 65_535;
    public static final int MAX_MESSAGE_BODY_BYTES = 1024 * 1024;
    public static final PacketCodec INSTANCE = new PacketCodec();

    public static final int NAT_BODY_HEADER_SIZE = 16;
    public static final int NAT_FLAG_END_STREAM = 1;
    public static final long MAX_STREAM_VALUE = 0xffff_ffffL;
    private static final TypeReference<Map<String, Object>> METADATA_TYPE = new TypeReference<>() {
    };

    private final Map<Byte, Class<? extends Packet>> packetTypeMap = new HashMap<>();

    private PacketCodec() {
        packetTypeMap.put(Command.LOGIN_REQUEST, LoginRequestPacket.class);
        packetTypeMap.put(Command.LOGIN_RESPONSE, LoginResponsePacket.class);
        packetTypeMap.put(Command.MESSAGE_REQUEST, MessageRequestPacket.class);
        packetTypeMap.put(Command.MESSAGE_RESPONSE, MessageResponsePacket.class);
        packetTypeMap.put(Command.LOGOUT_REQUEST, LogoutRequestPacket.class);
        packetTypeMap.put(Command.LOGOUT_RESPONSE, LogoutResponsePacket.class);
        packetTypeMap.put(Command.HEARTBEAT_REQUEST, HeartBeatRequestPacket.class);
        packetTypeMap.put(Command.HEARTBEAT_RESPONSE, HeartBeatResponsePacket.class);
        packetTypeMap.put(Command.NAT_MESSAGE, NatMessagePacket.class);
    }

    public void encode(ByteBuf byteBuf, Packet packet) throws ProtocolException {
        encode(byteBuf, packet, Serializer.COMPACT_BINARY);
    }

    public void encode(ByteBuf byteBuf, Packet packet, Serializer serializer) throws ProtocolException {
        if (packet == null) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "packet is required");
        }
        if (serializer == null || serializer.getSerializerAlgorithm() != SerializerAlgorithm.BIN) {
            throw new ProtocolException(
                    ProtocolException.Reason.UNSUPPORTED_SERIALIZER,
                    "control protocol v2 requires CompactBinary serializer");
        }
        byte command = packet.getCommand();
        if (!packetTypeMap.containsKey(command)) {
            throw new ProtocolException(ProtocolException.Reason.UNKNOWN_COMMAND, "unknown command: " + command);
        }

        byteBuf.writeInt(MAGIC_NUMBER);
        byteBuf.writeByte(PROTOCOL_VERSION);
        byteBuf.writeByte(SerializerAlgorithm.BIN);
        byteBuf.writeByte(command);

        if (command == Command.NAT_MESSAGE) {
            encodeNatMessage(byteBuf, (NatMessagePacket) packet);
            return;
        }

        byte[] body;
        try {
            body = serializer.serialize(packet);
        } catch (RuntimeException exception) {
            throw new ProtocolException(
                    ProtocolException.Reason.MALFORMED_BODY,
                    "cannot encode command " + command,
                    exception);
        }
        validateCommandBodyLength(command, body.length);
        byteBuf.writeInt(body.length);
        byteBuf.writeBytes(body);
    }

    public Packet decode(ByteBuf byteBuf) throws ProtocolException {
        if (byteBuf == null || byteBuf.readableBytes() < HEADER_SIZE) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_LENGTH, "frame is shorter than header");
        }
        int magic = byteBuf.readInt();
        if (magic != MAGIC_NUMBER) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_MAGIC, "invalid packet magic");
        }
        int version = byteBuf.readUnsignedByte();
        if (version != Byte.toUnsignedInt(PROTOCOL_VERSION)) {
            throw new ProtocolException(
                    ProtocolException.Reason.UNSUPPORTED_VERSION,
                    "unsupported protocol version: " + version);
        }
        int serializer = byteBuf.readUnsignedByte();
        if (serializer != SerializerAlgorithm.BIN) {
            throw new ProtocolException(
                    ProtocolException.Reason.UNSUPPORTED_SERIALIZER,
                    "unsupported serializer: " + serializer);
        }
        byte command = byteBuf.readByte();
        Class<? extends Packet> packetType = packetTypeMap.get(command);
        if (packetType == null) {
            throw new ProtocolException(ProtocolException.Reason.UNKNOWN_COMMAND, "unknown command: " + command);
        }
        int bodyLength = byteBuf.readInt();
        if (bodyLength < 0 || bodyLength != byteBuf.readableBytes()) {
            throw new ProtocolException(
                    ProtocolException.Reason.INVALID_LENGTH,
                    "declared body length does not match frame: " + bodyLength + "/" + byteBuf.readableBytes());
        }
        validateCommandBodyLength(command, bodyLength);

        try {
            if (command == Command.NAT_MESSAGE) {
                return decodeNatMessage(byteBuf, bodyLength);
            }
            byte[] body = new byte[bodyLength];
            byteBuf.readBytes(body);
            return Serializer.COMPACT_BINARY.deserialize(packetType, body);
        } catch (ProtocolException exception) {
            throw exception;
        } catch (RuntimeException exception) {
            throw new ProtocolException(
                    ProtocolException.Reason.MALFORMED_BODY,
                    "malformed body for command " + command,
                    exception);
        }
    }

    public Serializer getSerializer(byte serializeAlgorithm) {
        return serializeAlgorithm == SerializerAlgorithm.BIN ? Serializer.COMPACT_BINARY : null;
    }

    public Class<? extends Packet> getRequestType(byte command) {
        return packetTypeMap.get(command);
    }

    private void encodeNatMessage(ByteBuf byteBuf, NatMessagePacket packet) throws ProtocolException {
        NatMessageType type = packet.getNatMessageType();
        if (type == null) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "NAT message type is required");
        }
        byte[] encodedMetadata = encodeMetadata(packet.getMetaData());
        byte[] data = packet.getData() == null ? new byte[0] : packet.getData();
        validateNatSemantics(type, packet.getStreamId(), packet.getValue(), packet.getFlags(),
                encodedMetadata.length, data.length);
        long bodyLength = (long) NAT_BODY_HEADER_SIZE + encodedMetadata.length + data.length;
        if (bodyLength > Integer.MAX_VALUE) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_LENGTH, "NAT body is too large");
        }
        validateCommandBodyLength(Command.NAT_MESSAGE, (int) bodyLength);

        byteBuf.writeInt((int) bodyLength);
        byteBuf.writeByte(type.getCode());
        byteBuf.writeByte(packet.getFlags());
        byteBuf.writeShort(encodedMetadata.length);
        byteBuf.writeInt(packet.getStreamId());
        byteBuf.writeInt((int) packet.getValue());
        byteBuf.writeInt(data.length);
        byteBuf.writeBytes(encodedMetadata);
        byteBuf.writeBytes(data);
    }

    private Packet decodeNatMessage(ByteBuf byteBuf, int bodyLength) throws ProtocolException {
        if (bodyLength < NAT_BODY_HEADER_SIZE) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_LENGTH, "NAT body is shorter than header");
        }
        NatMessageType type = NatMessageType.fromWireId(byteBuf.readUnsignedByte());
        if (type == null) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "unknown NAT message type");
        }
        int flags = byteBuf.readUnsignedByte();
        if ((flags & ~NAT_FLAG_END_STREAM) != 0) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "unknown NAT flags: " + flags);
        }
        int metadataLength = byteBuf.readUnsignedShort();
        int streamId = byteBuf.readInt();
        long value = Integer.toUnsignedLong(byteBuf.readInt());
        int dataLength = byteBuf.readInt();
        if (dataLength < 0
                || metadataLength > MAX_NAT_METADATA_BYTES
                || metadataLength > byteBuf.readableBytes()
                || dataLength != byteBuf.readableBytes() - metadataLength) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_LENGTH, "invalid NAT metadata/data length");
        }
        byte[] metadataBytes = new byte[metadataLength];
        byteBuf.readBytes(metadataBytes);
        Map<String, Object> metadata = decodeMetadata(metadataBytes);
        byte[] data = dataLength == 0 ? null : new byte[dataLength];
        if (data != null) {
            byteBuf.readBytes(data);
        }
        if (byteBuf.isReadable()) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_LENGTH, "NAT body has trailing bytes");
        }
        validateNatSemantics(type, streamId, value, flags, metadataLength, dataLength);

        NatMessagePacket packet = new NatMessagePacket();
        packet.setNatMessageType(type);
        packet.setFlags(flags);
        packet.setStreamId(streamId);
        packet.setValue(value);
        packet.setMetaData(metadata);
        packet.setData(data);
        return packet;
    }

    private byte[] encodeMetadata(Map<String, Object> metadata) throws ProtocolException {
        if (metadata == null || metadata.isEmpty()) {
            return new byte[0];
        }
        byte[] encoded;
        try {
            encoded = Serializer.FASTJSON.serialize(metadata);
        } catch (RuntimeException exception) {
            throw new ProtocolException(
                    ProtocolException.Reason.MALFORMED_BODY,
                    "cannot encode NAT metadata",
                    exception);
        }
        if (encoded.length > MAX_NAT_METADATA_BYTES) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_LENGTH, "NAT metadata is too large");
        }
        return encoded;
    }

    private Map<String, Object> decodeMetadata(byte[] encoded) throws ProtocolException {
        if (encoded.length == 0) {
            return Map.of();
        }
        Map<String, Object> metadata = JsonUtil.bytesToObjectStrict(encoded, METADATA_TYPE);
        if (metadata == null) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "NAT metadata must be an object");
        }
        return metadata;
    }

    private void validateNatSemantics(NatMessageType type, int streamId, long value, int flags,
                                      int metadataLength, int dataLength) throws ProtocolException {
        if (value < 0 || value > MAX_STREAM_VALUE) {
            throw new ProtocolException(ProtocolException.Reason.INVALID_LENGTH, "NAT value is outside uint32");
        }
        boolean streamFrame = switch (type) {
            case OPEN, FIN, DATA, RST, WINDOW_UPDATE -> true;
            default -> false;
        };
        if (streamFrame == (streamId == 0)) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY,
                    streamFrame ? "stream frame requires a non-zero stream id"
                            : "connection frame requires stream id zero");
        }
        if (type != NatMessageType.DATA && flags != 0) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "flags are only valid on DATA");
        }
        if (type == NatMessageType.DATA && (metadataLength != 0 || value != 0)) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "DATA cannot carry metadata/value");
        }
        if (type == NatMessageType.FIN && (dataLength != 0 || flags != 0)) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY,
                    "FIN cannot carry binary data/flags");
        }
        if (type == NatMessageType.WINDOW_UPDATE
                && (metadataLength != 0 || dataLength != 0 || flags != 0)) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY,
                    "WINDOW_UPDATE cannot carry payload");
        }
        if (type == NatMessageType.WINDOW_UPDATE && value == 0) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "WINDOW_UPDATE credit must be positive");
        }
        if (type == NatMessageType.FIN && value != 0) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "FIN value must be zero");
        }
        if (type == NatMessageType.RST && dataLength != 0) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY, "RST cannot carry binary data");
        }
        if (!streamFrame && (value != 0 || flags != 0 || dataLength != 0)) {
            throw new ProtocolException(ProtocolException.Reason.MALFORMED_BODY,
                    "connection control frame cannot carry stream value/data");
        }
    }

    private void validateCommandBodyLength(byte command, int bodyLength) throws ProtocolException {
        int maximum = command == Command.LOGIN_REQUEST || command == Command.LOGIN_RESPONSE
                ? PRE_AUTH_MAX_FRAME_SIZE - HEADER_SIZE
                : command == Command.MESSAGE_REQUEST || command == Command.MESSAGE_RESPONSE
                ? MAX_MESSAGE_BODY_BYTES
                : Integer.MAX_VALUE;
        if (bodyLength < 0 || bodyLength > maximum) {
            throw new ProtocolException(
                    ProtocolException.Reason.INVALID_LENGTH,
                    "command " + command + " body exceeds limit: " + bodyLength + "/" + maximum);
        }
    }

}
